llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -595,9 +604,12 @@ void AIEPostRASchedStrategy::enterRegion(MachineBasicBlock *BB,
 
 void AIEPostRASchedStrategy::leaveRegion(const SUnit &ExitSU) {
   LLVM_DEBUG(dbgs() << "    << leaveRegion\n");
-  materializeMultiOpcodeInstrs();
 
   auto &BS = InterBlock.getBlockState(CurMBB);
+  if (BS.FixPoint.II) {
+    return;
+  }
```

Comment:
That state machine should be absolutely clear, since it is interpreted by a disparate set of call backs, and potentially represents the state of every basic block in the function.
I think that IsScheduled, IsPipelined, etc could all be represented by a single enumeration per block, which tell the different scheduler call backs tell what to do.  My gut feeling is that 'BuildRegionStructure' could be added as an initial non-scheduling state, which would probably tidy up some of the awkward locations of doing region management.
enterRegion would just administrate the regions, and enterBasicBlock could decide to iterate over the regions by itself.